compiler/backend/src/org/jetbrains/kotlin/codegen/inline/SMAPParser.kt

```diff
@@ -21,60 +21,64 @@ object SMAPParser {
+        for (line in iterator) {
+            when {
+                line == SMAP.LINE_SECTION -> break
+                line == SMAP.FILE_SECTION || line == SMAP.END || line.startsWith(SMAP.STRATA_SECTION) -> return null
             }
 
-            val indexAndFileInternalName = fileDeclaration.substringAfter("+ ").trim()
+            val indexAndFileInternalName = if (line.startsWith("+ ")) line.substring(2) else line
             val fileIndex = indexAndFileInternalName.substringBefore(' ').toInt()
             val fileName = indexAndFileInternalName.substringAfter(' ')
-            val path = iterator.next().trim()
+            val path = if (line.startsWith("+ ")) iterator.next() else fileName
             fileMappings[fileIndex] = FileMapping(fileName, path)
         }
 
-        for (lineMapping in iterator) {
-            // The stratum is terminated either by *E or another stratum.
-            if (lineMapping.trim().startsWith("*")) break
-            /*only simple mapping now*/
-            val targetSplit = lineMapping.indexOf(':')
-            val originalPart = lineMapping.substring(0, targetSplit)
-            val rangeSeparator = originalPart.indexOf(',').let { if (it < 0) targetSplit else it }
+        for (line in iterator) {
+            when {
+                line == SMAP.LINE_SECTION || line == SMAP.FILE_SECTION -> return null
+                line == SMAP.END || line.startsWith(SMAP.STRATA_SECTION) -> break
+            }
 
-            val fileSeparator = lineMapping.indexOf('#')
-            val originalIndex = originalPart.substring(0, fileSeparator).toInt()
-            val range = if (rangeSeparator == targetSplit) 1 else originalPart.substring(rangeSeparator + 1, targetSplit).toInt()
+            // <source>#<file>,<sourceRange>:<dest>,<destMultiplier>
+            val fileSeparator = line.indexOf('#')
+            if (fileSeparator < 0) return null
+            val destSeparator = line.indexOf(':', fileSeparator)
+            if (destSeparator < 0) return null
+            val sourceRangeSeparator = line.indexOf(',').let { if (it !in fileSeparator..destSeparator) destSeparator else it }
+            val destMultiplierSeparator = line.indexOf(',', destSeparator).let { if (it < 0) line.length else it }
 
-            val fileIndex = lineMapping.substring(fileSeparator + 1, rangeSeparator).toInt()
-            val targetIndex = lineMapping.substring(targetSplit + 1).toInt()
-            fileMappings[fileIndex]!!.mapNewInterval(originalIndex, targetIndex, range)
+            val file = fileMappings[line.substring(fileSeparator + 1, sourceRangeSeparator).toInt()] ?: return null
+            val source = line.substring(0, fileSeparator).toInt()
+            val dest = line.substring(destSeparator + 1, destMultiplierSeparator).toInt()
+            val range = when {
+                // These two fields have a different meaning, but for compatibility we treat them the same. See `SMAPBuilder`.
```

Comment:
What is migration plan for proper meaning?

Reply:
Could you also add navigation test with `// !LANGUAGE: +CorrectSourceMappingSyntax` (Please check that debugger support such directives)?

Reply:
Since CorrectSourceMappingSyntax is tagged as a 1.5 feature, the parser can be modified to handle the fields according to spec in 1.6, although I don't think this is particularly useful.

>Could you also add navigation tests with // !LANGUAGE: +CorrectSourceMappingSyntax (Please check that debugger support such directives)?

Done.

Reply:
NB: I'm mostly aware of debugger because he should distinguish old format from new one of proper ranging (but can't imagine any case where it would be major). Looks like class abi version could be used here